### PR TITLE
feat(policies): Add X-VLA

### PR DESCRIPTION
**Pull request by jadechoghari**
## What this does
feat(policies): Add X-VLA

X-VLA was proposed here: https://thu-air-dream.github.io/X-VLA/ and won Champion @ [AgiBot World Challenge @ IROS 2025](https://agibot-world.com/challenge)

X-VLA is a soft-prompted, flow-matching VLA framework that treats each hardware setup as a “task” and encodes it using a small set of learnable embeddings. These Soft Prompts capture embodiment and domain-specific variations, guiding the Transformer from the earliest stages of multimodal fusion. With this mechanism, X-VLA can reconcile diverse robot morphologies, data types, and sensor setups within a single unified architecture.

Libero also got updated to handle 1) different control mode, delta vs absolute - 2) you can now specify the max episode length, otherwise it will go to default depending on the task suite you choose

For finetuning / training
- VLM vision encoder: TRAINABLE
- VLM language encoder: TRAINABLE
- Policy transformer: TRAINABLE
- Soft prompts: TRAINABLE

We also introduce an auto mode that allows to train on any robot data out of the box (even though the max_action_dim of xvla-base is 20)

We provide the following checkpoints:
https://huggingface.co/lerobot/xvla-base (The base one pretrained on a large amount of robotics dataset)
https://huggingface.co/lerobot/xvla-folding (Success of 100% for 2 straight hours on folding cloth)
https://huggingface.co/lerobot/xvla-libero
https://huggingface.co/lerobot/xvla-google-robot
https://huggingface.co/lerobot/xvla-agibot-world
https://huggingface.co/lerobot/xvla-widowx

We also provide the cloth folding dataset of around 1.5k episodes of cloth folding: https://huggingface.co/datasets/lerobot/xvla-soft-fold

### CODE REVIEW

**Comment by 2toinf** [comment]
Further, I’d like to check X-VLA’s performance after post-training with the LeRobot pipeline. Does it match the officially reported results?

**Comment by jadechoghari** [comment]
Hello @2toinf yes this is standard in lerobot, we run a reproducibility check where we compare the expected logits from the preprocessor with the logits produced by lerobot implementation, and we also compare the expected logits of the produced actions with those from the original implementation
See: https://github.com/huggingface/lerobot/blob/171d50e85478537cfcae721845293b17beffd41d/tests/policies/xvla/test_xvla_original_vs_lerobot.py

Along with our Libero benchmark checker

File: pyproject.toml [hunk 1]
```diff
@@ -129,6 +129,7 @@ groot = [
     "ninja>=1.11.1,<2.0.0",
     "flash-attn>=2.5.9,<3.0.0 ; sys_platform != 'darwin'"
 ]
+xlva = ["lerobot[transformers-dep]"]
```
Comment (michel-aractingi) [inline pyproject.toml]: xlva typo...

File: src/lerobot/policies/xvla/modeling_xvla.py [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+                local_files_only=local_files_only,
+                revision=revision,
+                **kwargs,
+            )
+
+        model_id = str(pretrained_name_or_path)
+        instance = cls(config, **kwargs)
+        # step 2: locate model.safetensors
+        if os.path.isdir(model_id):
+            print("Loading weights from local directory")
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/modeling_xvla.py]: use logging.info instead of print

File: src/lerobot/policies/xvla/modeling_xvla.py [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+                    force_download=force_download,
+                    proxies=proxies,
+                    resume_download=resume_download,
+                    token=token,
+                    local_files_only=local_files_only,
+                )
+            except HfHubHTTPError as e:
+                raise FileNotFoundError(f"model.safetensors not found on the Hub at {model_id}") from e
+
+        print(f"Loading checkpoint from {model_file}")
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/modeling_xvla.py]: logging.info instead of print

File: src/lerobot/policies/xvla/modeling_xvla.py [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+        # step 3: load state dict
+        state_dict = safetensors.torch.load_file(model_file)
+        encoder_key = "model.vlm.language_model.model.encoder.embed_tokens.weight"
+        shared_key = "model.vlm.language_model.model.shared.weight"
+        if encoder_key in state_dict:
+            state_dict[shared_key] = state_dict[encoder_key]
+            # or deepcopy
+        # step 4: load into instance
+        instance.load_state_dict(state_dict, strict=True)
+        print("Loaded XVLA checkpoint")
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/modeling_xvla.py]: same here

File: src/lerobot/policies/xvla/processor_xvla.py [hunk 5]
```diff
@@ -0,0 +1,554 @@
...
+    This processor step adds a domain_id tensor to the complementary data,
+    which is used by XVLA to identify different robot embodiments or task domains.
+
+    Args:
+        domain_id: The domain ID to add (default: 3)
+        device: Device to place the domain_id tensor on (default: "cuda")
+    """
+
+    domain_id: int = 0
+    device: str = "cuda"
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/processor_xvla.py]: this seems hardcoded, can I run xvla if I don't have a gpu?
Can use DeviceProcessorStep since its already the next step in the pipeline?

File: src/lerobot/policies/xvla/__init__.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+# register the processor steps
+from lerobot.policies.xvla.processor_xvla import (
+    XVLAAddDomainIdProcessorStep,
+    XVLAImageNetNormalizeProcessorStep,
+    XVLAImageToFloatProcessorStep,
+)
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/__init__.py:6]: I don't think you need this files since these processors are used only in processor_xvla? 
Try removing this init file and check if the code will still work

File: src/lerobot/policies/xvla/action_hub.py [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+                pred[:, :, :real_dims],
+                target[:, :, :real_dims],
+            )
+            * self.JOINTS_SCALE
+        )
+
+        left_arm_loss = self.mse(pred[:, :, :6], target[:, :, :6])
+        right_arm_loss = self.mse(pred[:, :, 6:12], target[:, :, 6:12])
+
+        # is gripper continuous? not bce?
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/action_hub.py]: remove this comment?
 we decided continuous in the end?

File: src/lerobot/policies/xvla/processor_xvla.py [hunk 8]
```diff
@@ -0,0 +1,554 @@
...
+        comp = new_transition.get(TransitionKey.COMPLEMENTARY_DATA, {})
+        comp = {} if comp is None else comp.copy()
+
+        # Infer batch size from observation tensors
+        obs = new_transition.get(TransitionKey.OBSERVATION, {})
+        batch_size = 1
+        if obs:
+            for v in obs.values():
+                if isinstance(v, torch.Tensor):
+                    batch_size = v.shape[0]
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/processor_xvla.py:448]: you can probaby infer the device from obs? device = v.device

File: src/lerobot/policies/xvla/modeling_florence2.py [hunk 9]
```diff
@@ -0,0 +1,2868 @@
...
+            patch_size=config.patch_size,
+            patch_stride=config.patch_stride,
+            patch_padding=config.patch_padding,
+            patch_prenorm=config.patch_prenorm,
+            drop_path_rate=config.drop_path_rate,
+            window_size=config.window_size,
+        )
+
+
+if is_flash_attn_2_available():
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/modeling_florence2.py]: second condition if is_flash_attn...
Duplicated from line 56

File: src/lerobot/policies/xvla/modeling_florence2.py [hunk 10]
```diff
@@ -0,0 +1,2868 @@
...
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+"""
+
+
+@add_start_docstrings(
+    """The FLORENCE2 vision model without any head""",
+    FLORENCE2_START_DOCSTRING,
+)
+class Florence2VisionModel(Florence2PreTrainedModel):
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/modeling_florence2.py]: This class is unused? remove if true

File: src/lerobot/policies/xvla/modeling_florence2.py [hunk 11]
```diff
@@ -0,0 +1,2868 @@
...
+        else:
+            raise ValueError(f"invalid image shape {pixel_values.shape}")
+        return x
+
+
+@add_start_docstrings(
+    """The FLORENCE2 vision model with projection layer""",
+    FLORENCE2_START_DOCSTRING,
+)
+class Florence2VisionModelWithProjection(Florence2PreTrainedModel):
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/modeling_florence2.py]: Also this class is it used anywhere? remove if not used?

File: src/lerobot/policies/xvla/modeling_xvla.py [hunk 12]
```diff
@@ -0,0 +1,525 @@
...
+        return self._get_action_chunk(batch)
+
+    @torch.no_grad()
+    def select_action(self, batch: dict[str, Tensor], noise: Tensor | None = None) -> Tensor:  # noqa: ARG002
+        self.eval()
+        self._queues = populate_queues(self._queues, batch, exclude_keys=[ACTION])
+
+        if len(self._queues[ACTION]) == 0:
+            actions = self._get_action_chunk(batch)
+            self._queues[ACTION].extend(actions.transpose(0, 1)[: self.config.n_action_steps])
```
Comment (michel-aractingi) [inline src/lerobot/policies/xvla/modeling_xvla.py:414]: Verify that the actions are trimmed according to the requested action space as I had to manually trim it in the real robot test

**Review by michel-aractingi — COMMENTED** [review]
Overall great work Jade. The PR is very close to review

File: src/lerobot/policies/xvla/processor_xvla.py (hunk 5 above)
Comment (jadechoghari) [reply src/lerobot/policies/xvla/processor_xvla.py]: yes, i will remove this device, and keep device processor step after all custom steps

File: src/lerobot/policies/xvla/processor_xvla.py (hunk 8 above)
Comment (jadechoghari) [reply src/lerobot/policies/xvla/processor_xvla.py:448]: for the domain id, we can add any device, and this will be handled by device processor step, correct?

File: src/lerobot/policies/xvla/modeling_xvla.py (hunk 12 above)
Comment (jadechoghari) [reply src/lerobot/policies/xvla/modeling_xvla.py:414]: done

File: src/lerobot/policies/xvla/__init__.py (hunk 6 above)
Comment (jadechoghari) [reply src/lerobot/policies/xvla/__init__.py:6]: You need a way to register these, because if processors are defined inside a config (e.g., xvla_add_domain_id), they must be registered before being used

File: src/lerobot/policies/xvla/modeling_xvla.py (hunk 2 above)
Comment (jadechoghari) [reply src/lerobot/policies/xvla/modeling_xvla.py]: done

File: src/lerobot/policies/xvla/modeling_florence2.py (hunk 10 above)
Comment (jadechoghari) [reply src/lerobot/policies/xvla/modeling_florence2.py]: done

File: src/lerobot/policies/xvla/modeling_florence2.py (hunk 11 above)
Comment (jadechoghari) [reply src/lerobot/policies/xvla/modeling_florence2.py]: done

**Review by michel-aractingi — APPROVED** [review]
great work @jadechoghari ! 
Thanks to @2toinf and co for this method and the feedback

File: .github/workflows/full_tests.yml [hunk 13]
```diff
@@ -64,6 +64,13 @@ jobs:
           lfs: true
           persist-credentials: false
 
+      - name: Free disk space
+        run: |
+          sudo rm -rf /usr/share/dotnet
+          sudo rm -rf /usr/local/lib/android
+          sudo rm -rf /opt/ghc
+          docker system prune -af || true
+
```
Comment (jadechoghari) [inline .github/workflows/full_tests.yml]: @imstevenpmwork it is based on https://mathio28.medium.com/squeezing-disk-space-from-github-actions-runners-an-engineers-guide-d5fbe4443692 and frees 8-10 gbs of space for the ci (unused libraries)

Comment (imstevenpmwork) [reply .github/workflows/full_tests.yml]: Solved in: https://github.com/huggingface/lerobot/pull/2564

**Review by michel-aractingi — APPROVED** [review]

File: src/lerobot/policies/factory.py [hunk 14]
```diff
@@ -416,8 +416,7 @@ def make_policy(
             raise ValueError("env_cfg cannot be None when ds_meta is not provided")
         features = env_to_policy_features(env_cfg)
 
-    if not cfg.output_features:
```
Comment (carloscp3009) [inline src/lerobot/policies/factory.py:419]: Hello @jadechoghari , Can I ask you why this was deleted? I collected data with two arms, and I want to train only for one arm for some experiments, but the output_features is always overwriten `cfg.output_features = {key: ft for key, ft in features.items() if ft.type is FeatureType.ACTION} `

Comment (jadechoghari) [reply src/lerobot/policies/factory.py:419]: Hey @carloscp3009 thx for flagging this , could you share your dataset and policy
